### PR TITLE
[Algorithm] Support Circular Smooth Label (CSL, ECCV'20)

### PR DESCRIPTION
**Pull request by yangxue0827**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Support [Circular Smooth Label ](https://link.springer.com/chapter/10.1007/978-3-030-58598-3_40)(CSL, ECCV'20) for https://github.com/open-mmlab/mmrotate/issues/117. 
Special thanks to [liuyanyi](https://github.com/liuyanyi) for his help.

## Modification
- Add new head: [csl_rotated_retina_head](https://github.com/open-mmlab/mmrotate/blob/main/mmrotate/models/dense_heads/csl_rotated_retina_head.py)
- Add new loss: [smooth_focal_loss](https://github.com/open-mmlab/mmrotate/blob/main/mmrotate/models/losses/smooth_focal_loss.py)
- Add new coder: [angle_coder](https://github.com/open-mmlab/mmrotate/blob/main/mmrotate/core/bbox/coder/angle_coder.py)
- Add new configs: [rotated_retinanet_obb_csl_gaussian_r50_fpn_fp16_1x_dota_le90](https://github.com/open-mmlab/mmrotate/blob/main/configs/csl)
- Update documentations: [model_zoo_en](https://github.com/open-mmlab/mmrotate/blob/main/docs/en/model_zoo.md), [model_zoo_zh-cn](https://github.com/yangxue0827/mmrotate/blob/main/docs/zh_cn/model_zoo.md), [README](https://github.com/open-mmlab/mmrotate/blob/main/README.md), [README_zh-CN](https://github.com/open-mmlab/mmrotate/blob/main/README_zh-CN.md)

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/redet/redet_re50_refpn_1x_dota_ms_rr_le90.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-_base_ = ['./redet_re50_fpn_1x_dota_le90.py']
+_base_ = ['./redet_re50_refpn_1x_dota_le90.py']
 
 data_root = '/cluster/home/it_stu198/main/datasets/split_ms_dota1_0/'
 angle_version = 'le90'
```
Comment (jbwang1997) [inline configs/redet/redet_re50_refpn_1x_dota_ms_rr_le90.py:4]: Use the default data_root.

File: mmrotate/core/bbox/coder/angle_coder.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def encode(self, angle_targets):
+        pass
+
+    def decode(self, angle_preds):
+        pass
+
```
Comment (jbwang1997) [inline mmrotate/core/bbox/coder/angle_coder.py:125]: Here seems to be not complete. We can delete it first.

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+
+        loss_angle = self.loss_angle(
+            angle_cls,
+            angle_targets,
+            weight=angle_weights,
+            avg_factor=num_total_samples)
+
+        return loss_cls, loss_bbox, loss_angle
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'angle_classes'))
```
Comment (jbwang1997) [inline mmrotate/models/dense_heads/csl_rotated_retina_head.py:156]: angle_classes is not appear in `loss` function inputs.

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+        if with_nms:
+            det_bboxes, det_labels = multiclass_nms_rotated(
+                mlvl_bboxes, mlvl_scores, cfg.score_thr, cfg.nms,
+                cfg.max_per_img)
+            return det_bboxes, det_labels
+        else:
+            return mlvl_bboxes, mlvl_scores
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'angle_classes'))
+    def get_bboxes(self,
```
Comment (jbwang1997) [inline mmrotate/models/dense_heads/csl_rotated_retina_head.py:468]: `angle_classes` doesn't apear in inputs.

File: configs/redet/redet_re50_refpn_1x_dota_ms_rr_le90.py (hunk 1 above)
Comment (jbwang1997) [reply configs/redet/redet_re50_refpn_1x_dota_ms_rr_le90.py:4]: It seems a personal root. I mean we can use `data/split_ms_dota1_0` as default.

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+                                                    angle_cls_list,
+                                                    mlvl_anchors, img_shape,
+                                                    scale_factor, cfg, rescale)
+            else:
+                proposals = self._get_bboxes_single(cls_score_list,
+                                                    bbox_pred_list,
+                                                    angle_cls_list,
+                                                    mlvl_anchors, img_shape,
+                                                    scale_factor, cfg, rescale,
+                                                    with_nms)
```
[the remark this answers is not in the document]
Comment (yangxue0827) [reply mmrotate/models/dense_heads/csl_rotated_retina_head.py:562]: Yes, I modified it based on other heads. This issue doesn't seem suitable to be addressed on this PR.

File: configs/csl/rotated_retinanet_obb_csl_gaussian_r50_fpn_fp16_1x_dota_le90.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+_base_ = \
+    ['../rotated_retinanet/rotated_retinanet_obb_r50_fpn_fp16_1x_dota_le90.py']
+
```
[the remark this answers is not in the document]
Comment (yangxue0827) [reply configs/csl/rotated_retinanet_obb_csl_gaussian_r50_fpn_fp16_1x_dota_le90.py:3]: What do you mean?

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/models/dense_heads/csl_rotated_retina_head.py:562]: I get it.

File: mmrotate/core/anchor/anchor_generator.py [hunk 7]
```diff
@@ -43,6 +44,47 @@ def single_level_grid_priors(self,
         return anchors
 
 
+@ROTATED_ANCHOR_GENERATORS.register_module()
+class FakeRotatedAnchorGenerator(AnchorGenerator):
+    """Fake rotate anchor generator for 2D anchor-based detectors."""
+
+    def __init__(self, angle_version='oc', **kwargs):
```
Comment (zytx121) [inline mmrotate/core/anchor/anchor_generator.py:51]: AnchorGenetator's parameters must be explicitly passed in the configuration file, otherwise, users may mistakenly adopt the default `oc` horizontal box when using `le90` or `le135`. It's best to give a configuration file to show users how to use it. If it is not used at present, I suggest deleting this part of the code first.

File: mmrotate/core/bbox/coder/__init__.py [hunk 8]
```diff
@@ -1,10 +1,11 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from .angle_coder import CSLCoder
```
Comment (zytx121) [inline mmrotate/core/bbox/coder/__init__.py:2]: The name of `angle_coder` is not clear. We'd better rename it to `csl_coder` or another name.

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+
+from mmrotate.core import build_bbox_coder, multiclass_nms_rotated
+from ... import obb2hbb, rotated_anchor_inside_flags
+from ..builder import ROTATED_HEADS, build_loss
+from .rotated_retina_head import RotatedRetinaHead
+
+
+@ROTATED_HEADS.register_module()
+class CSLRRetinaHead(RotatedRetinaHead):
+
```
Comment (zytx121) [inline mmrotate/models/dense_heads/csl_rotated_retina_head.py:16]: Docstring is required here.

File: mmrotate/core/bbox/coder/__init__.py (hunk 8 above)
Comment (yangxue0827) [reply mmrotate/core/bbox/coder/__init__.py:2]: `angle_coder` will continue to add the `DCLCoder`, so it is not suitable to be renamed to `csl_coder`. Actually, I think the word angle has made the meaning quite clear.

**Review by zytx121 — APPROVED** [review]
LGTM

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py (hunk 3 above)
Comment (zytx121) [reply mmrotate/models/dense_heads/csl_rotated_retina_head.py]: It should be `angle_classes` or `angle_clses`?

File: mmrotate/models/dense_heads/csl_rotated_retina_head.py (hunk 4 above)
Comment (zytx121) [reply mmrotate/models/dense_heads/csl_rotated_retina_head.py]: It looks like a typo.

File: configs/csl/metafile.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    Training Techniques:
+      - SGD with Momentum
+      - Weight Decay
+    Training Resources: 1x GeForce GTX 2080 Ti
+    Architecture:
+      - ResNet
+  Paper:
+    URL: None
+    Title: 'Arbitrary-Oriented Object Detection with Circular Smooth Label'
+  README: configs/csl/README.md
```
Comment (jbwang1997) [inline configs/csl/metafile.yml:14]: Paper URL and README are incorrect. And need a Code field.

File: configs/csl/rotated_retinanet_obb_csl_gaussian_r50_fpn_fp16_1x_dota_le90.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbwang1997) [reply configs/csl/rotated_retinanet_obb_csl_gaussian_r50_fpn_fp16_1x_dota_le90.py:3]: We can directly use `_base_ = '../rotated_retinanet/rotated_retinanet_obb_r50_fpn_fp16_1x_dota_le90.py'`

**Review by jbwang1997 — APPROVED** [review]
Have a little metafile.yml error. Others look good to me.

File: configs/csl/metafile.yml (hunk 10 above)
Comment (zytx121) [reply configs/csl/metafile.yml]: We previously referenced MMPose's metafile.yml setting, so there is no code field. Maybe, we could propose another PR later to modify it uniformly.

**Comment by zytx121** [comment]
The model weights and logs have been uploaded.

File: mmrotate/core/bbox/coder/angle_coder.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        # empty label
+        smooth_label = torch.zeros_like(angle_targets).repeat(
+            1, self.coding_len)
+        angle_targets_deg = (angle_targets_deg +
+                             self.angle_offset) / self.omega
+        # Float to Int
+        angle_targets_long = angle_targets_deg.long()
+
+        if self.window == 'pulse':
+            radius_range = angle_targets_long
```
Comment (liuyanyi) [inline mmrotate/core/bbox/coder/angle_coder.py:63]:
```suggestion
            radius_range = angle_targets_long % self.coding_len
```
an error occurred when use omega=8 window=pulse

Comment (yangxue0827) [reply mmrotate/core/bbox/coder/angle_coder.py]: done
